tests/testthat/test-class-Megastudy.R

```diff
@@ -0,0 +1,547 @@
+  expect_error(StudySpecificVocabulary(studyIdColumnName=c(), study='a', variableSpec=VariableSpec(entityId='sample',variableId='species'), vocabulary=c('species1','species2','species3')))
+  expect_error(StudySpecificVocabulary(studyIdColumnName=character(), study='a', variableSpec=VariableSpec(entityId='sample',variableId='species'), vocabulary=c('species1','species2','species3')))
+  expect_error(StudySpecificVocabulary(studyIdColumnName=NA_character_, study='a', variableSpec=VariableSpec(entityId='sample',variableId='species'), vocabulary=c('species1','species2','species3')))
+
+  # errs if no var spec provided for study vocab
+  expect_error(StudySpecificVocabulary(studyIdColumnName='study.id', study='a', vocabulary=c('species1','species2','species3')))
+  expect_error(StudySpecificVocabulary(studyIdColumnName='study.id', study='a', variableSpec=NA, vocabulary=c('species1','species2','species3')))
+  expect_error(StudySpecificVocabulary(studyIdColumnName='study.id', study='a', variableSpec=NULL, vocabulary=c('species1','species2','species3')))
+  # expect_error(StudySpecificVocabulary(studyIdColumnName='study.id', study='a', variableSpec=VariableSpec(entityId='', variableId=''), vocabulary=c('species1','species2','species3')))
+})
+
+# TODO this could go in its own file maybe
+test_that("imputeZeroes method is sane", {
+  m <- Megastudy(data=megastudyDT,
+                 ancestorIdColumns=c('study.id', 'collection.id', 'sample.id'),
+                 studySpecificVocabularies=StudySpecificVocabulariesByVariableList(S4Vectors::SimpleList(speciesVocabs)))
+
+  # case where neither study nor collection vars in the plot
+  variables <- new("VariableMetadataList", SimpleList(
+    new("VariableMetadata",
+      variableClass = new("VariableClass", value = 'native'),
+      variableSpec = new("VariableSpec", variableId = 'species', entityId = 'sample'),
+      plotReference = new("PlotReference", value = 'xAxis'),
+      dataType = new("DataType", value = 'STRING'),
+      dataShape = new("DataShape", value = 'CATEGORICAL'),
+      weightingVariableSpec = VariableSpec(variableId='specimen_count',entityId='sample'),
+      hasStudyDependentVocabulary = TRUE),
+    new("VariableMetadata",
+      variableClass = new("VariableClass", value = 'native'),
+      variableSpec = new("VariableSpec", variableId = 'specimen_count', entityId = 'sample'),
+      plotReference = new("PlotReference", value = 'yAxis'),
+      dataType = new("DataType", value = 'NUMBER'),
+      dataShape = new("DataShape", value = 'CONTINUOUS'))
+  ))
+
+  imputedDT <- getDTWithImputedZeroes(m, variables)
+  # result has the columns needed to build a plot, based on variables AND the correct number of rows/ zeroes
+  # TODO lol its just possible this fxn shouldnt remove cols but my brain hurts enough already 
+  expect_equal(all(c("sample.species","sample.specimen_count") %in% names(imputedDT)), TRUE)
+  expect_equal(nrow(imputedDT), 12)
```

Comment:
I calculate 8 if only considering the sample.species variable?

Collections `a.1, a.2, b.1, b.2` should each be seeing both species, so 4x2 = 8?

Reply:
ah but look again at the vocabs i defined, there are three species per study here.

Reply:
Ah!  Cool, I get it now.